### ISSUE TITLE
Inconsistent Newline Behavior in Converted Domino Database

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1033]
While discussing #675 with @piotrzarzycki21 in the Wednesday meeting, we found some inconsistent behavior for the converted `PerformanceStats_BingTest.nsf` database (this is not on GitHub, but I shared it previously).

There are many cells in the HostConfigSnapshot form that display a command and a corresponding field, on separate lines.  However, some of these cells are converted to a single paragraph in Domino Visual Editor, while others are converted to two paragraphs.  The single paragraph case seems like an unexpected configuration to me.  I converted the database again today with the latest NSFConverter version, and I got the same results:

![image](https://user-images.githubusercontent.com/24532731/169404753-1e5e0c8a-6b69-46f0-a602-f479c568e190.png)

It appears like this in Domino Designer:
![image](https://user-images.githubusercontent.com/24532731/169404909-7a8aea40-463d-440f-9903-06706cd71a61.png)

I see that Domino Visual Editor converts this back to the original format.  This satisfied the tests we were doing for the round-trip tests, but it doesn't seem consistent with the component structure.

![image](https://user-images.githubusercontent.com/24532731/169404796-6c5e29fe-b95a-4216-8f6c-80b0d6d02086.png)

**Comment by JoelProminic** [comment]
Here is the DXL for these two cells:

```
<tablerow>
    <tablecell>
        <pardef id='10' keepwithnext='true' keeptogether='true'/>
        <par def='10'><run><font name='monospace'/>hostname</run></par>
    </tablecell>
    <tablecell>
        <pardef id='11' keepwithnext='true' keeptogether='true'/>
        <par def="11">
          <run><font name="monospace"/>$ hostname<break/></run>
          <run>
            <font name="monospace"/>
            <field type="text" kind="editable" name="HostnameOutput"/>
          </run>
          <compositedata type="98" prevtype="65418" nexttype="222" afterparcount="7" containertype="65418" aftercontainercount="1" afterbegincount="3">
                Yg4BAIQAAAAAAAAAAAA=
                </compositedata>
          <run>
            <font name="monospace"/>
          </run>
        </par>
    </tablecell>
</tablerow>
<tablerow>
    <tablecell>
        <par def='10'><run><font name='monospace'/>uname</run></par>
    </tablecell>
    <tablecell>
        <par def='11'>
            <run><font name='monospace'/>$ uname -a</run>
        </par>
        <par def='11'>
            <run><font name='monospace'/></run>
            <run>
                <font name='monospace'/>
                <field type='text' kind='editable' name='UnameOutput'/>
            </run>
            <compositedata type='98' prevtype='65418' nexttype='222' afterparcount='6' containertype='65418' aftercontainercount='1' afterbegincount='3'>
            Yg4BAIQAAAAAAAAAAAA=
            </compositedata>
            <run><font name='monospace'/></run>
        </par>
    </tablecell>
</tablerow>
```

I see the `hostname` example has a `<break/>` after the command, while the `uname` example has multiple `<par>` entries.  So it seems that this is inconsistent on the Domino side as well.  I'm not sure why I ended up with this discrepancy.


Here is the XML in Domino Visual Editor (Code Tab):

```
              <tablerow>
                <tablecell hpostion="left">
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="system" style="normal" name="monospace"/>hostname
                    </run>
                  </par>
                </tablecell>
                <tablecell hpostion="left">
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="system" style="normal" name="monospace"/>$ hostname
                    </run>
                    <break/>
                    <run>
                      <font color="black" size="10pt" style="normal" name="sans-serif" pitch="variable" familyid="Roman" truetype="true"/><field listdisplayseparator="newline" listinputseparators="comma semicolon newline" useappletinbrowser="false" allowtabout="false" defaultfocus="false" storelocally="false" value="HostnameOutput" type="text" kind="editable" computeaftervalidation="false" allowmultivalues="false" width="106pt" height="30pt" bgcolor="#ffffff" name="HostnameOutput"/>
                    </run>
                  </par>
                </tablecell>
              </tablerow>
              <tablerow>
                <tablecell hpostion="left">
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="system" style="normal" name="monospace"/>uname
                    </run>
                  </par>
                </tablecell>
                <tablecell hpostion="left">
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="system" style="normal" name="monospace"/>$ uname -a
                    </run>
                  </par>
                  <par def="1307" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
                    <run>
                      <font color="black" size="10pt" style="normal" name="sans-serif" pitch="variable" familyid="10" truetype="true"/><field listdisplayseparator="newline" listinputseparators="comma semicolon newline" useappletinbrowser="false" allowtabout="false" defaultfocus="false" storelocally="false" value="UnameOutput" type="text" kind="editable" computeaftervalidation="false" allowmultivalues="false" width="100pt" height="30pt" bgcolor="#ffffff" name="UnameOutput"/>
                    </run>
                  </par>
                </tablecell>
              </tablerow>
```

I see that this also has a `<break/>` entry, but this doesn't have a representation in the mockup editor.

So, it seems that this behavior is consistent with the DXL, but it is confusing on the Visual Editor side.  I see a couple ways we could go with this:
- Have NSF converter (or the intermediate XML converter) split a paragraph in the multiple paragraphs when encountering a `<break/>`.  This would ideally result in cleaner intermediate XML, but we may also find some unexpected edge case for `<break/>` in the DXL.
- Add a clear representation of the `<break/>` in the Mockup Editor.  This would give us more flexibility on the NSF conversion, but it feels uintuitive in the mockup editor (that said, the current behavior is also unintuitive when the paragraphs are not displayed vertically).

**Comment by JoelProminic** [comment]
@feather812002, we would like to boost the priority on this task.  This technically not a blocker for #675, but it is a major distraction as we are trying to convert the Performance Stats database, which has a lot of `<break/>` cases.

I would like to resolve this by splitting DominoParagraphs with `<break/>` (i.e. `<par>...part1...<break/>...part2...<par>`) into two paragraphs:

```
<par>
    ...part1...
</par>
<par>
    ...part2...
</par>
```

This should produce the same result in the generated Domino database, and it will remove the need to handle the break in the Mockup Editor or converters.

One thing we may need to watch for is formatting - any formatting options may need to be copied to both paragraphs (if it is not applied to the individual components already).

You can test this with the PerformanceStats_BingTest.nsf database - see the examples above.

**Comment by piotrzarzycki21** [comment]
> @feather812002, we would like to boost the priority on this task. This technically not a blocker for #675, but it is a major distraction as we are trying to convert the Performance Stats database, which has a lot of `<break/>` cases.
> 
> I would like to resolve this by splitting DominoParagraphs with `<break/>` (i.e. `<par>...part1...<break/>...part2...<par>`) into two paragraphs:
> 
> ```
> <par>
>     ...part1...
> </par>
> <par>
>     ...part2...
> </par>
> ```
> 
> This should produce the same result in the generated Domino database, and it will remove the need to handle the break in the Mockup Editor or converters.
> 
> One thing we may need to watch for is formatting - any formatting options may need to be copied to both paragraphs (if it is not applied to the individual components already).
> 
> You can test this with the PerformanceStats_BingTest.nsf database - see the examples above.

@feather812002 This part has priority right now, so please your next task should be apply fix for that.

**Comment by feather812002** [comment]
**I can't understand why the dxl format will affect other formats convert.
Dxl format only work for Domino .**

Yeah, the **<break/>** not show a new line in the Domino Visual Editor.
So, I will fix this in the Domino Visual Editor ,let it display as a new line ,

but not change the <break> in the dxl , because  the <break> is Official tag from Domino , so why do we need change a Official tag to a custom tags??
https://help.hcltechsw.com/dom_designer/10.0.1/basic/H_BREAK_ELEMENT_XML.html
This is a very bad requirement.

If you want change the dxl format to a strange format, I can't guarantee that the dxl you convert back matches the original  domino interface, because you want to replace the correct tag <break> with another tag <par>, but require <par> to display the same as <break> ,it's out of the question.
@JoelProminic @JustinProminic @piotrzarzycki21 

Of course if you guys stick to this practice, I have no problem, but I can't guarantee that the converted format matches the original format.
Please confirm above comment .

**Comment by JoelProminic** [comment]
Sorry for the confusion, I see I only reported the original DXL and the generated DXL.  The intermediate XML for the hostname/uname example in the description looks like this:

```
            <Row tablabel="">
              <Column altbgcolor="" borderwidth="" colorstyle="" columnspan="" rowspan="" valign="">
                <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="10" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" familyid="" hide="" hidewhen="" isBreak="false" isNewLine="false" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype="">hostname</Label>
                    </run>
                  </Paragraph>
                </Div>
              </Column>
              <Column altbgcolor="" borderwidth="" colorstyle="" columnspan="" rowspan="" valign="">
                <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" familyid="" hide="" hidewhen="" isBreak="false" isNewLine="false" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype="">$ hostname</Label>
                      <Label color="system" familyid="" hide="" hidewhen="" isBreak="true" isNewLine="true" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype=""/>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                      <Field allowmultivalues="" formula="" height="30" hide="" hidewhen="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="HostnameOutput" recalcchoices="" recalonchange="" required="false" taborder="" type="text" value="HostnameOutput" width="106"/>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                    </run>
                  </Paragraph>
                </Div>
              </Column>
            </Row>
            <Row tablabel="">
              <Column altbgcolor="" borderwidth="" colorstyle="" columnspan="" rowspan="" valign="">
                <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="10" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" familyid="" hide="" hidewhen="" isBreak="false" isNewLine="false" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype="">uname</Label>
                    </run>
                  </Paragraph>
                </Div>
              </Column>
              <Column altbgcolor="" borderwidth="" colorstyle="" columnspan="" rowspan="" valign="">
                <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" familyid="" hide="" hidewhen="" isBreak="false" isNewLine="false" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype="">$ uname -a</Label>
                    </run>
                  </Paragraph>
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                      <Field allowmultivalues="" formula="" height="30" hide="" hidewhen="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="UnameOutput" recalcchoices="" recalonchange="" required="false" taborder="" type="text" value="UnameOutput" width="100"/>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                    </run>
                  </Paragraph>
                </Div>
              </Column>
            </Row>
```


So, `<break />` becomes `<Label color="system" familyid="" hide="" hidewhen="" isBreak="true" isNewLine="true" isUrlLink="false" name="monospace" pitch="" size="" style="normal" truetype=""/>` in the intermediate XML, and then it is converted back to `<break />` in the generated DXL.

However, this doesn't really change the fundamental issue.  I can see a corresponding label in the mockup and Organizer, but I have no way to change the properties that control whether this is a newline.  We could add properties for this, but I feel they would be unintuitive in the mockup.

**Comment by feather812002** [comment]
@JoelProminic  do you know why  it can't change the fundamental issue?
Because Designer IDE work with Text editor mode , but Moonshine-IDE work with flex layout mode .
These are two completely different editing modes, I memory we already discuss this before we start the Domino Visual Editor , but in last , you and other guys still insist on work convert the Text edit mode to Flex layout editor mode .

**The fundamental issue is you require me try let the text editor mode change to Flex layout editor mode .
but it can convert back to text editor mode and keep it UI style not change !**

You can input a newline with press "Enter" key when you work with Designer IDE,
You input the "Enter" key will get a newline , you input a space , you will get a space from the text editor , 
 but can you do this with any Flex layout editor ?

From my knowledge and experience I only see the flex layout work with "left,center,right" , or "top,middle,bottom", some similar layout  way, now our way is to simulate the UI of the text editor through flex layout, so it must bring some different .

So do you have any suggestion we can fix the  fundamental issue? 
I will replace the `<break>` to `<paragraph>`, and I sure it will broken the source Notes UI.
If this you want ,let us back if we got some trouble on this.

**Comment by feather812002** [comment]
@JoelProminic @JustinProminic 
I check the email, I think you will found a email on  02/28/2020
Moonshine IDE RichTextEdirtor Demo.

So , if we work on this mode , I can try to let the Moonshine IDE editor UI 100% match the Designer editor UI, 
But at least , you decide is we still work on the Moonshine-IDE Flex(Html) layout mode .

Primeface or Flex UI layout base on the Html layout , so for current visual editor mode ,it is fine and enough , 
But for Designer IDE layout mode , it have big different , I already try to suggestion this before 2 years ago.

But unfortunately, we still choose the current way to do Designer editor mode.
This is why it alway get some different and it always have  fundamental issue.

If you or other guys have more better idea can fix this, just let me know , I will follow the idea fix it ASAP.

Thanks.

**Comment by feather812002** [comment]
```
 <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" fontName="monospace" hide="" isBreak="false" isNewLine="false" isUrlLink="false" style="normal">$ hostname</Label>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                    </run>
                  </Paragraph>
                  <Paragraph>
                    <run hide="">
                      <font name="monospace"/>
                      <Field fontName="monospace" height="30" hide="" hidewhen="" htmlClass="" htmlId="" htmlOther="" htmlStyle="" htmlTitle="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="HostnameOutput" required="false" type="text" value="HostnameOutput" width="106"/>
                    </run>
                  </Paragraph>
                </Div>
```

OKay, Now, I already change the  intermediate XML follow on above format
![image](https://user-images.githubusercontent.com/8543391/172380049-6ee71001-2831-4a18-84ab-e322ce681044.png)
![image](https://user-images.githubusercontent.com/8543391/172380501-026e0fa9-af7f-4a92-a0d1-5da0bb59b067.png)


If this is you want ,just let us keep this ,and go ahead.

**Comment by feather812002** [comment]
It break the other` "<break>"` format , source UI as follow:
![image](https://user-images.githubusercontent.com/8543391/172381831-0c6fdab9-1c59-4b9a-9184-367e383f0fb9.png)
Now, the latest convert from break to paragraph:
![image](https://user-images.githubusercontent.com/8543391/172381999-f19dc78e-7f87-4d48-9d06-bcb470f8852c.png)

**Comment by feather812002** [comment]
The middle xml change to :  
```
<Paragraph>
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">I would run a verbose installation of the boot files:</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -v</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">That will output messages saying whether work needed to be done or if things</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">were up-to-date.</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also do a forced install of the required boot files,</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">including the MBR block with:</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -Mfv</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also tell if you're using UEFI boot with:</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # prtconf -v /devices | grep -A1 efi-version</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">If there is any output then you're booted under EFI. In that case, the ESPs</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">(EFI System Partitions) need synchronising too, but the above bootadm</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">install-bootloader command takes care of that.</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">Andy</Label>
                      </run>
                    </Paragraph>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
                    <Paragraph/>
```

So , this is why I ask you on last night , how can I decide the `<Paragraph/>` will add the child element or it just as a empty a  `<Paragraph/>`?

I no idea about this at now .

**Comment by feather812002** [comment]
After discuss with @JoelProminic  in today , he still decide we continue keep this way .
I am fine , he suggestion , if we get the 2 `<break>` , we convert it as a `<paragraph>`,
if it only one `<break>`, we split it as into two `<paragraph>`.

I already point out , it must broken other some case in this way ,
for example , if user input 3 `<break>`, from UI ,it should be  two new line space and one return "\n",

but follow above logic , it should convert one return , and convert other two `<break>` to one new line space.
So , it must incorrect UI after converted and generated again.

I add more discuss details in here:

From my view  : 
 1. the `<break>`is a placeholder, I convert it to a label, it can precise positioning position, exactly how many breaks are converted, and keep exactly the same as the original file. it don't need care the the preceding child element and the following child element
 2. the` <Paragraph/>`is a container ,it will convert to a `<par>`
     https://help.hcltechsw.com/dom_designer/10.0.1/basic/H_PARAGRAPH_ELEMENT_XML.html
      <par> it is a close tag ,it is a container , it must consider what child element inside it ,then we can get a correct UI

I will continue work on the this follow Joel's suggestion .
Let us to see if we can get a correct UI again .

**Comment by JoelProminic** [comment]
The demo @feather812002 mentioned is pretty basic, but it gives an idea of how this could work:

![image](https://user-images.githubusercontent.com/24532731/172495727-3a3e388a-fae5-4c01-9260-1fe3be6042ba.png)

We can discuss this in the group meeting tomorrow, but I doubt there will be much interest in rebuilding Visual Editor as a rich text editor.  One of the original ideas behind the Visual Editor project was defining the forms with more structure, so that it would (in theory) be easier to generate code for it in a variety of languages.  This is definitely causing some problems for NSFConverter, but we have demonstrated that we can round-trip a reasonably complicated database with this design.

Another alternative solution we could consider for this issue would be to allow Labels to have newline characters, which could be represented as `&#10;` or even `<break />` in the intermediate XML.   However, I think the solution I proposed makes more sense for managing the structure and layout of the form in the current editor.

**Comment by feather812002** [comment]
@JoelProminic 
Okay ,Joel ,follow your require , I done some update :  
1. intermedia xml from NSFConverter
```
<Section hide="" title="Additional information from Andy Fiddaman" titleColor="teal">
                    <Label color="system" hide="" isBreak="false" isNewLine="true" isUrlLink="false" style="normal"/>
                    <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" style="bold" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" fontStyle="bold" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">; b) be assured that the master boot record, boot loader, etc. on the other</Label>
                        <Label color="system" familyid="30" fontName="Consolas" fontStyle="bold" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">; two disks will actually work?</Label>
                      </run>
                    </Paragraph>
                    <Paragraph>
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">were up-to-date.</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">including the MBR block with:</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">(EFI System Partitions) need synchronising too, but the above bootadm</Label>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">install-bootloader command takes care of that.</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">I would run a verbose installation of the boot files:</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -v</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">That will output messages saying whether work needed to be done or if things</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also do a forced install of the required boot files,</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -Mfv</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also tell if you're using UEFI boot with:</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # prtconf -v /devices | grep -A1 efi-version</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">If there is any output then you're booted under EFI. In that case, the ESPs</Label>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">Andy</Label>
                    </Paragraph>
                  </Section>
```

2. Convert to dxl  : 
 ```
  <section expanded="false" showastext="false">
                    <sectiontitle color="teal">
                      <font style="normal"/><text>Additional information from Andy Fiddaman</text>
                    </sectiontitle>
                    <par isNewLine="true" paragraph="true">
                      <run>
                        <font color="system" size="12pt" style="normal"/></run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" style="bold" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>; b) be assured that the master boot record, boot loader, etc. on the other
                      </run>
                      <run>
                        <font color="system" style="bold" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>; two disks will actually work?
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>were up-to-date.
                      </run>
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>including the MBR block with:
                      </run>
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>(EFI System Partitions) need synchronising too, but the above bootadm
                      </run>
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>install-bootloader command takes care of that.
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>I would run a verbose installation of the boot files:
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>		 # bootadm install-bootloader -v
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>That will output messages saying whether work needed to be done or if things
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>You can also do a forced install of the required boot files,
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>		 # bootadm install-bootloader -Mfv
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>You can also tell if you're using UEFI boot with:
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>		 # prtconf -v /devices | grep -A1 efi-version
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>If there is any output then you're booted under EFI. In that case, the ESPs
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph" isNewLine="true">
                      <run isNewLine="true">
                        <font color="system" size="12pt" style="normal"/>
                      </run>
                    </par>
                    <par def="1089" paragraph="true" dominotype="paragraph">
                      <run>
                        <font color="system" name="Consolas" pitch="fixed" truetype="true" familyid="30"/>Andy
                      </run>
                    </par>
                  </section>
```

3. Generated notes UI:
   
![image](https://user-images.githubusercontent.com/8543391/172863378-c9ca4422-23b9-4aec-bd3a-9407320c0860.png)

4. source notes UI:
![image](https://user-images.githubusercontent.com/8543391/172863471-f81b7992-ae70-47f1-9f5c-f40cfa14f3fc.png)


It broken the text order and the new line order , like I type on above , the `<par>` tag must need close it , it not a Placeholder, when follow your logic, the trouble not is convert the `<break>` to empty` <par>`, the trouble is  when we insert a new empty` <par>`, we must move out  the other child element from the old `<par>`, then insert a new empty `<par>`, then follow other child element , so some times, it will broken text order and other child element order.

Other , in this case we only consider two `<break>`, do you think about if user input 3, or 4 `<break>`, what the solution for this ?

**Comment by JoelProminic** [comment]
Looking at your generated output, it seems that when there is only a single `<break />`, the following line was moved to the original paragraph instead of being put in a new, separate paragraph.  I only see this behavior in the Section, though - "$ hostname" is still fine.

I manually updated the intermediate XML like I expected.  I included some comments to describe what I changed.  I copied the `<run>` and `<font>` elements from the first non-bold paragraph:

```
                  <Section hide="" title="Additional information from Andy Fiddaman" titleColor="teal">
                    <Label color="system" hide="" isBreak="false" isNewLine="true" isUrlLink="false" style="normal"/>
                    <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" style="bold" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" fontStyle="bold" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">; b) be assured that the master boot record, boot loader, etc. on the other</Label>
                      </run>
                    </Paragraph>  <!-- JCA:  split the first paragraph -->
                    <Paragraph class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" wrap="false">
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" style="bold" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" fontStyle="bold" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">; two disks will actually work?</Label>
                      </run>
                    </Paragraph>
                    <!--
                    <Paragraph>
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                      </run>
                    </Paragraph>
                    - This paragraph had incorrectly positioned labels.  Once I removed them, the paragraph was no longer necessary -->
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">I would run a verbose installation of the boot files:</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -v</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">That will output messages saying whether work needed to be done or if things</Label>
                      </run>
                    </Paragraph>
                    <Paragraph>
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">were up-to-date.</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also do a forced install of the required boot files,</Label>
                      </run>
                    </Paragraph>
                    <Paragraph> <!-- moved from first paragraph -->
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">including the MBR block with:</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # bootadm install-bootloader -Mfv</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">You can also tell if you're using UEFI boot with:</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">		 # prtconf -v /devices | grep -A1 efi-version</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">If there is any output then you're booted under EFI. In that case, the ESPs</Label>
                      </run>
                    </Paragraph>
                    <Paragraph> <!-- moved from first paragraph -->
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">(EFI System Partitions) need synchronising too, but the above bootadm</Label>
                      </run>
                    </Paragraph>
                    <Paragraph> <!-- moved from first paragraph -->
                      <run hide="">
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">install-bootloader command takes care of that.</Label>
                      </run>
                    </Paragraph>
                    <Paragraph isNewLine="true"/>
                    <Paragraph>
                      <run hide=""> <!-- Missing <run> and <font> in split paragraphs. Was this intentional? -->
                        <font familyid="30" name="Consolas" pitch="fixed" truetype="true"/>
                        <Label color="system" familyid="30" fontName="Consolas" hide="" isBreak="false" isNewLine="false" isUrlLink="false" pitch="fixed" style="normal" truetype="true">Andy</Label>
                      </run>
                    </Paragraph>
                  </Section>
                  
```

This result looks close to the original, except for these issues (I think this might be an old issue):
- The field appears after the section rather than before it.
- There is an extra newline after the section header.

**Comment by feather812002** [comment]
@JoelProminic @JustinProminic 
In today's discuss , I suggestion, the  intermediate XML keep follow Joel's format, but I will add some additional property for the  intermediate XML, it not effect any other convert , it just some additional property for some tag.
and I can sure we get the correct Notes client again .

In this case, we can got follow result:
1. make sure the intermediate XML follow Joel's intermediate XML  format ,
2. make sure the converted and generated dxl file ,more and more match old source format.
3. make sure we get the notes client UI more and more match old notes client UI .

But unfortunately, Joel still think we must need replace the `<break>` with `<par>`, not only in the  intermediate XML, but also in the dxl file , I don't know why ?
It look like we try to give up a official tag , and replace it by other tag ,even in the dxl file , 
Joel also think the `<break>` exactly equal to empty `<par>`,
so we can full replace it whatever in the  intermediate XML  and  the dxl file of domino both.

So, in next work , I will work follow Joel's design and to see if we can let it work for any case .

I also hope all will work fine , otherwise , we have to back to here to check this again .

Thanks.

**Comment by JoelProminic** [comment]
@feather812002,  I reviewed [this code](https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/e296bea914963de8fd9d6a062329080e99b24413/dxl-to-visual-editor-converter/src/main/java/net/prominic/dxl/converter/service/MoonshineDominoXmlConverter.java#L562) after our discussion today.

I think this code is way more complicated than it needs to be.  I think the discussion about `<break/>` led you in the wrong direction for this.

I tried my own, simplified implementation for this.   When it finds a break, it will:
1. Create a new sibling of the current paragraph
2. Move all elements after the break to the new paragraph
3. Remove the break node

Some other decisions I made:
- I set `isNewLine="true"` for both the original and new paragraph, to make sure they would display as newlines if the other children were removed.  This could be done in a smarter way if necessary
- I did not try to duplicate the Paragraph attributes, or the `font` or `run` nodes.

The result turned out better than I expected.
![image](https://user-images.githubusercontent.com/24532731/173706107-46ac201d-a6c5-4054-9fc6-9be790b6e478.png)

The only remaining problems I is that there are a missing newline before the Section, and an extra newline at the beginning of the session.  I think these are old problems, though.

I committed my changes as `replaceAllBreaks`.  I kept your old code as `replaceBreakOld`.  Take a look to see if there is something I missed.

**Comment by piotrzarzycki21** [comment]
Discussion about new line behaviour -> https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1040#issuecomment-1196440802

**Comment by piotrzarzycki21** [comment]
@JoelProminic Ok I [did change default behaviour](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1041#issuecomment-1149698349) of cell in DominoTable - it is now Vertical, so I believe you should made the changes on converter sight so I could get properly exported containers of cells. It should be change as my comment here -> https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1040#issuecomment-1196440802

**Comment by JoelProminic** [comment]
Yeah, I see this is waiting on changes in #1041.   I pinged @feather812002 on this.

**Comment by JoelProminic** [comment]
I moved the remaining #1041 issues to a separate task. We can create a new issue if this does not resolve the problem.

FYI, I am getting errors while testing the PerformanceStats_BingTest database with the Royale converter now, so I can't confirm the current behavior.  We'll revisit this conversion in 3.3.4.
